runtime/pubsub/utils/utils.go

```diff
@@ -0,0 +1,187 @@
+package utils
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"strconv"
+	"time"
+)
+
+// MarshalFields creates a map[string]string of fields in `msg` tagged with `tag`. The name of the tag
+// will be used as map key, and values are converted to strings using fmt.Sprintf. Pointers will be dereferenced
+// and ignored if nil. Only basic types (bool, numeric, string) and pointers to those types are supported fields.
+// Fields without a tag will not be marshalled. `msg` must be a struct or pointer to a struct
+func MarshalFields[T any](msg T, tag string) (map[string]string, error) {
```

Comment:
As a short-term future todo; we should look to add caching to these two marshal functions for types (as json.Marshal does)

As a long-term todo; we should add code generation to types to remove all need for reflection and looping here

Reply:
Added caching for the unmarshalling for now. The marshalling stuff is pretty much using the same encoder for everything.